File: app/static/js/i18n.js
/* Лёгкий i18n-слой без сборки: словари ru/uk встроены в страницу сервером
   (window.__I18N__, см. index.html) — так переводы применяются синхронно,
   до того как остальные скрипты успеют отрисовать динамический контент.
   Переключение языка сохраняется в localStorage и применяется мгновенно,
   без перезагрузки страницы (см. setLanguage/langchange). */
(function () {
  const STORAGE_KEY = "dashboard_lang";
  const DICTS = window.__I18N__ || { ru: {}, uk: {} };

  let currentLang = localStorage.getItem(STORAGE_KEY) || "ru";
  if (!DICTS[currentLang]) currentLang = "ru";

  function t(key, vars) {
    const dict = DICTS[currentLang] || {};
    let text = dict[key];
    if (text === undefined) text = (DICTS.ru || {})[key];
    if (text === undefined) return key;
    if (vars) {
      Object.keys(vars).forEach((k) => {
        text = text.replace(new RegExp(`\\{${k}\\}`, "g"), vars[k]);
      });
    }
    return text;
  }

  function applyTranslations(root) {
    root = root || document;
    root.querySelectorAll("[data-i18n]").forEach((el) => {
      el.textContent = t(el.getAttribute("data-i18n"));
    });
    root.querySelectorAll("[data-i18n-html]").forEach((el) => {
      el.innerHTML = t(el.getAttribute("data-i18n-html"));
    });
    root.querySelectorAll("[data-i18n-placeholder]").forEach((el) => {
      el.setAttribute("placeholder", t(el.getAttribute("data-i18n-placeholder")));
    });
    root.querySelectorAll("[data-i18n-title]").forEach((el) => {
      el.setAttribute("title", t(el.getAttribute("data-i18n-title")));
    });
  }

  function setLanguage(lang) {
    if (!DICTS[lang]) return;
    currentLang = lang;
    localStorage.setItem(STORAGE_KEY, lang);
    document.documentElement.lang = lang === "uk" ? "uk" : "ru";
    applyTranslations();
    document.dispatchEvent(new CustomEvent("langchange", { detail: { lang } }));
  }

  // Автоматически прокидываем текущий язык интерфейса на бэкенд (заголовок X-Lang),
  // не трогая ни один из ~40 существующих fetch()-вызовов по всем вкладкам.
  const nativeFetch = window.fetch.bind(window);
  window.fetch = function (input, init) {
    init = init || {};
    let headers;
    if (input instanceof Request) {
      headers = new Headers(input.headers);
    } else {
      headers = new Headers(init.headers || {});
    }
    headers.set("X-Lang", currentLang);
    init.headers = headers;
    return nativeFetch(input, init);
  };

  function locale() {
    return currentLang === "uk" ? "uk-UA" : "ru-RU";
  }

  document.documentElement.lang = currentLang === "uk" ? "uk" : "ru";
  window.I18N = { t, applyTranslations, setLanguage, getLang: () => currentLang, locale };
  applyTranslations();

  // Переключатель РУС/УКР в шапке — общий для index.html и onboarding.html, поэтому живёт
  // здесь, а не в app.js (который тянет за собой логику вкладок, ненужную мастеру настройки).
  document.addEventListener("DOMContentLoaded", () => {
    const langButtons = document.querySelectorAll(".lang-btn");
    function syncLangButtons() {
      langButtons.forEach((b) => b.classList.toggle("active", b.dataset.lang === currentLang));
    }
    langButtons.forEach((btn) => {
      btn.addEventListener("click", () => {
        setLanguage(btn.dataset.lang);
        syncLangButtons();
      });
    });
    syncLangButtons();
  });
})();

File: app/i18n.py
"""
Переводы интерфейса (ru/uk) — единый источник: app/static/lang/{ru,uk}.json.
Одни и те же ключи используются и на фронтенде (через data-i18n / window.I18N.t()),
и на бэкенде — для текстов, которые формирует сервер (сообщения статусов/ошибок).
Бэкенд определяет язык по заголовку X-Lang, который i18n.js прозрачно
проставляет на каждый fetch() (см. app/static/js/i18n.js).
"""
import json
import os

from flask import request

LANG_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "static", "lang")
DEFAULT_LANG = "ru"
SUPPORTED_LANGS = ("ru", "uk")

_cache = {}


def _load(lang: str) -> dict:
    if lang not in _cache:
        path = os.path.join(LANG_DIR, f"{lang}.json")
        with open(path, "r", encoding="utf-8") as f:
            _cache[lang] = json.load(f)
    return _cache[lang]


def load_translations() -> dict:
    """Оба словаря целиком — встраиваются в index.html при рендере (см. app/__init__.py)."""
    return {lang: _load(lang) for lang in SUPPORTED_LANGS}


def current_lang() -> str:
    lang = (request.headers.get("X-Lang") or "").strip()
    return lang if lang in SUPPORTED_LANGS else DEFAULT_LANG


def t(key: str, **template_vars) -> str:
    lang = current_lang()
    text = _load(lang).get(key) or _load(DEFAULT_LANG).get(key) or key
    for k, v in template_vars.items():
        text = text.replace(f"{{{k}}}", str(v))
    return text
